src/include/migraphx/op/unique.hpp

```diff
@@ -0,0 +1,340 @@
+    }
+
+    // Default: none. Range: [-rank, rank-1]
+    static constexpr int64_t axis_none = std::numeric_limits<int64_t>::max();
+    int64_t axis                       = axis_none;
+
+    // Default: 1= sorted. 0 = unsorted.
+    int64_t sorted = 1;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.axis, "axis"), f(self.sorted, "sorted"));
+    }
+
+    std::string name() const { return "unique"; }
+
+    shape compute_shape(std::vector<shape> inputs) const
+    {
+        check_shapes{inputs, *this}.has(1).standard();
+
+        auto& sh_x         = inputs[0];
+        auto lens_x        = sh_x.lens();
+        size_t dim_x       = sh_x.ndim();
+        size_t max_uniq_ct = sh_x.elements();
+
+        int64_t t_axis = axis;
+        if(t_axis != axis_none)
+        {
+            t_axis = migraphx::tune_axis(dim_x, t_axis, name());
+            if(t_axis != 0)
+                MIGRAPHX_THROW("Unique: Only supports axis = 0 or None");
+            max_uniq_ct /= sh_x.strides()[t_axis];
+        }
+
+        // min = 1 unique element; max = full dimension along the axis
+        // The three outputted Indices are just 1-D:
+        std::vector<shape::dynamic_dimension> d_out{{1, max_uniq_ct}};
+        shape sh_idx{shape::int64_type, {d_out}};
+
+        // Unique elements themselves aren't necessarily 1-D.. fix its output shape:
+        for(size_t idx = 1; t_axis != axis_none && t_axis + idx < dim_x; idx++)
+            d_out.push_back({lens_x[t_axis + idx], lens_x[t_axis + idx]});
+        shape sh_y = {sh_x.type(), {d_out}};
+
+        return {{sh_y, sh_idx, sh_idx, sh_idx}};
+    }
+
+    argument compute(const dyn_output& dyn_out, std::vector<argument> args) const
+    {
+        size_t uniq_ct     = 0;
+        auto sh_x          = args.front().get_shape();
+        size_t dim_x       = sh_x.ndim();
+        auto lens_x        = sh_x.lens();
+        shape output_shape = dyn_out.computed_shape;
+        auto vec_ss        = output_shape.sub_shapes();
+        auto x_ct          = sh_x.elements();
+        shape sh_y         = {vec_ss[0].type(), {x_ct}};
+        shape sh_idx       = {vec_ss[1].type(), {x_ct}};
+        shape sh_x_idx     = {vec_ss[1].type(), {x_ct}};
+
+        argument res_y{sh_y};
+        argument res_y_idx{sh_idx};
+        argument res_x_rev_idx{sh_idx};
+        argument res_y_ct_idx{sh_idx};
+
+        std::vector<size_t> out_y_idx;
+        std::vector<size_t> out_x_rev_idx;
+        std::vector<size_t> out_y_ct;
+
+        // If axis is not none, for >1D tensors, we have to consider
+        // then, the uniqueness of chunks of sub-tensors: a subsequence of built-ins..
+        // For a built-in type, chunk_sz is of course = 1
+        size_t chunk_sz = 1;
+        int64_t t_axis  = axis;
+        if(axis != axis_none)
+        {
+            t_axis   = tune_axis(dim_x, t_axis, name());
+            chunk_sz = sh_x.strides()[t_axis];
```

Comment:
This is cumulatively adding the chunk size (of all the indices on the right of the t_axis). Hence the strides().

Reply:
If its not standard shape then the stride may not be equal to the accumulated dimensions. You should probably use `std::accumulate` to compute this instead or maybe `sh_x.elements() / sh_x.lens()[t_axis]`.

Reply:
This operator is standard shape. Hence strides() is proper to use -- rather than to effectively repeat its recalculation via accumulate etc. Lens() is not useful here. Thanks.

Reply:
> This operator is standard shape.

Ref operators should not require standard shape, which means this operator needs to work with inputs that are non standard, which means you can not rely on the strides here.

Reply:
Is there any operator that could input to this Unique Operator with a non-standard shape()? Thanks.

Reply:
Many operators will produce non-standard shape like slice,step,transpose,broadcast,etc, but many more will propagate the permutation in memory as well. 

We dont want to require standard shape for ref operators because when we do graph optimizations it can lead to the shape changing indirectly downstream to a non-standard shape which will lead to an error. We had many of these issues previously when reshape required a standard shape(like #1496, #1579, #2136) which is why we reimplemented reshape to handle nonstandard inputs(#1677, #2099). 

Furthermore, ref operators are evaluated during constant propagation which may have standard shape inputs, so they should be written robustly to handle all memory layouts. 

For backends like gpu, and cpu, they can require standard shape, and we ensure it will be a standard shape if an operator needs it. After lowering we dont do high-level graph rewrites(like algebraic simplifications) that would change the layout in memory.

Reply:
OK. Thanks. I will accumulate those results from the lens, and not use strides().

Reply:
Strides or accumulate have been replaced by a simpler logic, since we allow only the axis=0, so not relevant any more.